### PR TITLE
Remove Ctype.cyclic_abbrev

### PR DESCRIPTION
**Pull request by stedolan**
While looking at the implementation of `[@@unboxed]`, I got lost down a rabbit-hole of type expansion. I got very confused by `Ctype.cyclic_abbrev` in particular: it's used to detect and reject aliases like `type t = t`, but since it runs in the `temp_env` it can't detect cycles such as `type t = u and u = t`.

Those are detected by a better check later in `Typedecl.check_well_founded`, which seems to subsume `Ctype.cyclic_abbrev`. I can't find anything that `Ctype.cyclic_abbrev` will catch that won't be caught by `check_well_founded`, so this patch deletes the former. (I've verified that various cyclic abbrevs are still rejected, and the testsuite still passes)

### CODE REVIEW

**Review by lpw25 — COMMENTED** [review]
I also cannot see how `cyclic_abbrev` can be doing anything useful. I'm "seconding" this PR by removing the `Draft` marker, but I think I'd like approval from @garrigue before merging in case we're both missing something.

**Comment by garrigue** [comment]
You would probably have to ask @vouillon for why we have two separate checks.
My superficial understanding is that, by running earlier, `cyclic_abbrev` shall be able to protect us from a certain category of problem during the typing of declarations.
Of course, as we have seen in #9314, we cannot currently completely protect against cycles, so there is no real panacea.

While this test is indeed not sufficient to remove all cycles, it is hard to tell whether there exists a situation where it allows to produce an error rather than having the compiler loop forever (or until exhaustion). The testsuite is hardly complete in that respect.

Removing `cyclic_abbrev` might still help clarify the situation.

**Comment by xavierleroy** [comment]
> My superficial understanding is that, by running earlier, cyclic_abbrev shall be able to protect us from a certain category of problem during the typing of declarations.

My vague recollection of the early days of recursive modules was that there were many bugs of the kind "the typechecker loops because an illegal cyclic abbreviation was created and not checked yet".  

I am confident that `Typedecl.check_well_founded` subsumes `Ctype.cyclic_abbrev`, but are we sure the former is run early enough?  And why is the latter run earlier than the former?  (if I understand @garrigue's comment correctly).

**Comment by garrigue** [comment]
Looking at the code, the difference of timing between the two checks is very small: `Ctype.cyclic_abbrev` runs at the end of `Typedecl.transl_declaration`, which itself is called only in one place, inside `Typedecl.transl_type_decl` a mere few lines before `check_well_founded_manifest`. In between, there are only updates to the environment and calls to generalize, which should not rely on well-foundedness (they do no expansion).
So the only real difference between the two seems to be the environment used, so that `cyclic_abbrev` does not expend mutual recursive types, while `check_well_founded` does a much more thorough job, and in particular `check_well_founded_manifest` seems to completely subsume `cyclic_abbrev`. The only risk is that `check_well_founded_manifest` itself would loop, but I don't see how `cyclic_abbrev` would protect against that.

This said, `check_well_founded` is much more complicated than `cyclic_abbrev`, so the only argument I would have to keep it is that it at least guarantees that some simple cases are caught early.

Note that `check_well_founded` has changed a lot over the years, so it might be that originally it was doing things less carefully than it does now.

**Comment by lpw25** [comment]
Those responses match my own when I looked at this: you can see why you might need a less precise check to run before it was safe to run `check_well_founded` but, looking at the details of `cyclic_abbrev` and when it is run, that does not appear to be the case here.

I'm going to go ahead and mark this as approved based on Jacques and I both thinking that the code is unnecessary. (I would just merge it, but I guess it should have a Changes entry first).

**Review by lpw25 — APPROVED** [review]

**Comment by vouillon** [comment]
It is well possible that `cyclic_abbrev` was attempting to do what `check_well_founded` (which was added later by @garrigue) is doing properly

**Comment by gasche** [comment]
> (I would just merge it, but I guess it should have a Changes entry first).

Well enough waiting for the Changes entry, I'll merge now.

**Pull request merged**
